File: questions/q10/eg1.c
// 10. To check the linked list is pallindrome. 

#include<stdio.h>
#include<stdlib.h>
struct Node
{
int data;
struct Node * next;
};
struct Node * start=NULL;
struct Node * createNode(int data)
{
struct Node *t;
t=(struct Node *)malloc(sizeof(struct Node));
t->data=data;
t->next=NULL;
return t;
}
void createLinkedList()
{
struct Node *n1,*n2,*n3,*n4,*n5;
n1=createNode(10);
n2=createNode(20);
n3=createNode(30);
n4=createNode(20);
n5=createNode(10);
start=n1;
n1->next=n2;
n2->next=n3;
n3->next=n4;
n4->next=n5;
}
void releaseStack(struct Node *b)
{
struct Node *t;
t=(struct Node *)malloc(sizeof(struct Node));
t=b;
b=b->next;
free(t);
}
int isPallindrome(struct Node *b)
{
struct Node *p1,*p2,*top,*t;
top=NULL;
int count=1;
if(b==NULL) return 0;
if(b->next==NULL) return 1;
p2=b;
p1=b->next;
// push a node on stack
t=createNode(p2->data);
t->next=top;
top=t;

while(p1!=NULL)
{
p2=p2->next;
//push p2(a node) on stack
t=createNode(p2->data);
t->next=top;
top=t;

if(p1->next==NULL)
{
count+=1;
break;
}
p1=p1->next->next;
count+=2;
}
if(count%2==0)
{
// pop a node from stack
t=top;
top=top->next;
free(t);
}
while(p2!=NULL)
{
// compare p2->data with top->data if not equal
if(p2->data!=top->data)
{
releaseStack(top);
return 0;
}
// pop a node from stack
t=top;
top=top->next;
free(t);
p2=p2->next;
}
return 1;
}
int main()
{
createLinkedList();
if(isPallindrome(start)) printf("Pallindrome");
else printf("Not a pallindrome");
return 0;
}

File: questions/q10/eg1.java
// Find the given linked list is palindrome or not
import java.util.*;
public class eg1
{
static class Node
{
int data;
Node next;
Node()
{
this.data=0;
this.next=null;
}
Node(int d)
{
this.data=d;
this.next=null;
}
}
static Node start=new Node();
public static void createLinkedList()
{
Node n1=new Node(10);
Node n2=new Node(20);
Node n3=new Node(30);
Node n4=new Node(20);
Node n5=new Node(20);
start=n1;
n1.next=n2;
n2.next=n3;
n3.next=n4;
n4.next=n5;
}
public static boolean isPallindrome(Node b)
{
Node p1=b;
boolean isPalindrome=true;
Stack<Integer> stack=new Stack<Integer>();
while(p1!=null)
{
stack.push(p1.data);
p1=p1.next;
}
while(b!=null)
{
int i=stack.pop();
if(b.data==i) isPalindrome=true;
else
{
isPalindrome=false;
break;
}
b=b.next;
}
return isPalindrome;
}
public static void main(String args[])
{
createLinkedList();
System.out.println("Is Pallindrome : "+isPallindrome(start));
}
}